### PR TITLE
Refactors page API

### LINKED ISSUE #137458: How should Cupertino back gesture interact with onPopPage
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

Currently the Cupertino back gesture assumes its pop will always remove the route, however it missed the case where onPopPage can still veto the pop. Right now, this will cause the page to hang.

The question is what should they do in this case. I can think several ways forward.
1. onPopPage can veto the pop, then we have to fix the cupertino back swipe to not freeze the screen if it is veto'd
2. page based route will disable cupertino back swipe.
3. cupertino backswipe can't be veto by onPopPage.

It seems to me (2) or (3) is closer to the current cupertino back swipe behavior, but (2) may be too aggressive. If we want to implement (3) we will need a new API in Navigator to still give customer a chance to clean up their page list.

### Actual results

fix the issue.

### Logs

<details>
<summary>Logs</summary>

```console

<!-- Paste your logs here -->

```

</details>

### Flutter Doctor output

<details>
<summary>Doctor output</summary>

```console

<!-- Paste your output here -->

```

</details>

### PR DESCRIPTION
**Pull request by chunhtai**
fixes https://github.com/flutter/flutter/issues/137458

Chagnes:
1. Navigator.pop will always pop page based route
2. add a onDidRemovePage callback to replace onPopPage
3. Page.canPop and Page.onPopInvoked mirrors the PopScope, but in Page class.

migration guide https://github.com/flutter/website/pull/10523

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by chunhtai** [comment]
I will add test later if this approach looks good

File: packages/flutter/lib/src/widgets/navigator.dart [hunk 1]
```diff
@@ -636,6 +662,28 @@ abstract class Page<T> extends RouteSettings {
...
+  final PopInvokedCallback onPopInvoked;
+
+  /// When false, blocks the associated route from being popped.
+  ///
+  /// This includes the root route, where upon popping, the Flutter app would
+  /// exit.
+  ///
+  /// If there are also [PopScope] widgets appear in a route's widget subtree,
+  /// each of their `canPop` must be `true` in order for the route to be
+  /// able to pop.
```
Comment (justinmc) [inline packages/flutter/lib/src/widgets/navigator.dart:684]: Maybe add this clarification: "each of their `canPop` must be `true`, in addition to this `canPop`, in order for the route to be able to pop."

File: packages/flutter/lib/src/widgets/navigator.dart [hunk 2]
```diff
@@ -3528,37 +3606,39 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin, Res
...
-        if (widget.pages.isEmpty) {
-          FlutterError.reportError(
-            FlutterErrorDetails(
-              exception: FlutterError(
-                'The Navigator.pages must not be empty to use the '
-                'Navigator.pages API',
-              ),
-              library: 'widget library',
-              stack: StackTrace.current,
+  bool _debugCheckPageApiParameters() {
```
Comment (justinmc) [inline packages/flutter/lib/src/widgets/navigator.dart:3609]: Good call pulling this out into a function.

File: packages/flutter/lib/src/widgets/navigator.dart [hunk 3]
```diff
@@ -1508,8 +1561,27 @@ class Navigator extends StatefulWidget {
...
+  /// (e.g. the value returned from a dialog).
+  ///
+  /// This callback is responsible for calling [Route.didPop] and returning
+  /// whether this pop is successful.
+  ///
+  /// The [Navigator] widget should be rebuilt with a [pages] list that does not
+  /// contain the [Page] for the given [Route]. The next time the [pages] list
+  /// is updated, if the [Page] corresponding to this [Route] is still present,
+  /// it will be interpreted as a new route to display.
+  final DidRemovePageCallback? onDidRemovePage;
```
Comment (justinmc) [inline packages/flutter/lib/src/widgets/navigator.dart:1583]: The name sounds good to me.

File: packages/flutter/lib/src/widgets/navigator.dart [hunk 4]
```diff
@@ -636,6 +662,28 @@ abstract class Page<T> extends RouteSettings {
   ///    Flutter.
   final String? restorationId;
 
+  /// Called after a pop on the associated route was handled.
```
Comment (justinmc) [inline packages/flutter/lib/src/widgets/navigator.dart:665]: Nit: Is there anything that could be deduplicated with a macro in these docs?  Here and for `canPop`.

**Review by justinmc — COMMENTED** [review]
I've been away from predictive back for awhile, but this approach looks good to me.  It seems a lot simpler with no onPopPage and the same `canPop` and `onPopInvoked` API everywhere.  Especially because it works with system backs out of the box as users will expect.

Thank you for refactoring this!

**Comment by chunhtai** [comment]
@justinmc There is one thing I am not too sure about the popscope. Is there a way for PopScope to know what result the original pop used? For example if the pop is generated by Navigator.maybePop('some text'), if you want to launch a confirmation dialog in popScope, there isn't seem to be a way to get the string as pop result if user click confirm

File: packages/flutter/test/widgets/navigator_test.dart [hunk 5]
```diff
@@ -2890,8 +2890,8 @@ void main() {
         error.toStringDeep(),
         equalsIgnoringHashCodes(
           'FlutterError\n'
-          '   The Navigator.onPopPage must be provided to use the\n'
-          '   Navigator.pages API\n',
+          '   The one of onDidRemovePage and onPopPage must be provided to use\n'
```
Comment (goderbauer) [inline packages/flutter/test/widgets/navigator_test.dart:2893]:
```suggestion
          '   Either onDidRemovePage or onPopPage must be provided to use\n'
```

File: packages/flutter/lib/src/widgets/navigator.dart [hunk 6]
```diff
@@ -5229,14 +5287,10 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin, Res
 
     // TODO(justinmc): When the deprecated willPop method is removed, delete
     // this code and use only popDisposition, below.
-    final RoutePopDisposition willPopDisposition = await lastEntry.route.willPop();
-    if (!mounted) {
-      // Forget about this pop, we were disposed in the meantime.
-      return true;
-    }
```
[the remark this answers is not in the document]
Comment (chunhtai) [reply packages/flutter/lib/src/widgets/navigator.dart]: reverted

**Comment by btrautmann** [comment]
Hi 👋 Any idea when this might land? 🙏

**Comment by justinmc** [comment]
> @justinmc There is one thing I am not too sure about the popscope. Is there a way for PopScope to know what result the original pop used? For example if the pop is generated by Navigator.maybePop('some text'), if you want to launch a confirmation dialog in popScope, there isn't seem to be a way to get the string as pop result if user click confirm

I pasted an example below that hopefully covers what you mean.

I don't think you can get the result of the pop that triggered onPopInvoked from inside of onPopInvoked. I'm not sure we want to enable that. I think that's the same as it was with WillPopScope.  See the example, where I show the dialog from inside of onPopInvoked and receive the result there.

<details>

<summary>Example</summary>

```dart
import 'package:flutter/material.dart';

void main() => runApp(const NavigatorPopHandlerApp());

class NavigatorPopHandlerApp extends StatelessWidget {
  const NavigatorPopHandlerApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      initialRoute: '/home',
      routes: <String, WidgetBuilder>{
        '/home': (BuildContext context) => const _HomePage(),
        '/two': (BuildContext context) => const _PageTwo(),
      },
    );
  }
}

class _HomePage extends StatefulWidget {
  const _HomePage();

  @override
  State<_HomePage> createState() => _HomePageState();
}

class _HomePageState extends State<_HomePage> {

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            const Text('Page One'),
            TextButton(
              onPressed: () {
                Navigator.of(context).pushNamed('/two');
              },
              child: const Text('Next page'),
            ),
          ],
        ),
      ),
    );
  }
}

class _PageTwo extends StatefulWidget {
  const _PageTwo();

  @override
  State<_PageTwo> createState() => _PageTwoState();
}

class _PageTwoState extends State<_PageTwo> {
  Future<bool?> _showBackDialog() {
    return showDialog<bool>(
      context: context,
      builder: (BuildContext context) {
        return AlertDialog(
          title: const Text('Are you sure?'),
          content: const Text(
            'Are you sure you want to leave this page?',
          ),
          actions: <Widget>[
            TextButton(
              style: TextButton.styleFrom(
                textStyle: Theme.of(context).textTheme.labelLarge,
              ),
              child: const Text('Nevermind'),
              onPressed: () {
                Navigator.pop(context, false);
              },
            ),
            TextButton(
              style: TextButton.styleFrom(
                textStyle: Theme.of(context).textTheme.labelLarge,
              ),
              child: const Text('Leave'),
              onPressed: () {
                Navigator.pop(context, true);
              },
            ),
          ],
        );
      },
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            const Text('Page Two'),
            PopScope(
              canPop: false,
              onPopInvoked: (bool didPop) async {
                if (didPop) {
                  return;
                }
                final NavigatorState navigator = Navigator.of(context);
                final bool? shouldPop = await _showBackDialog();
                if (shouldPop ?? false) {
                  navigator.pop();
                }
              },
              child: TextButton(
                onPressed: () {
                  _showBackDialog();
                },
                child: const Text('Go back'),
              ),
            ),
          ],
        ),
      ),
    );
  }
}
```

</details>

**Comment by justinmc** [comment]
@btrautmann The author is on leave until April, but maybe I can take this over if people need it.

**Comment by justinmc** [comment]
@chunhtai Thanks, I see what you mean. Indeed I don't think it's possible to do that right now without your PR. I think we should put both PRs on hold for now until you get back.

**Comment by gd46** [comment]
Is there a timeline for this PR, and will this fix GoRouter to work with PopScope? I see a lot of open issues, and haven't found a version of GoRouter that works with PopScope on iOS or Android for our implementation.

**Comment by chunhtai** [comment]
This is ready for review @justinmc @goderbauer

**Comment by johnninja** [comment]
I've just encountered this issue, and it coincided with an update; I'm waiting online. lol.

File: packages/flutter/lib/src/widgets/navigator.dart [hunk 7]
```diff
@@ -610,6 +633,15 @@ class RouteSettings {
 /// The type argument `T` is the corresponding [Route]'s return type, as
 /// used by [Route.currentResult], [Route.popped], and [Route.didPop].
 ///
+/// The [canPop] and [onPopInvoked] are used for intercepting pops.
```
Comment (justinmc) [inline packages/flutter/lib/src/widgets/navigator.dart:636]: "The [canPop] and [onPopInvoked]" => "The [canPop] and [onPopInvoked] parameters"

File: packages/flutter/lib/src/widgets/navigator.dart [hunk 8]
```diff
@@ -610,6 +633,15 @@ class RouteSettings {
...
+///
+/// {@tool dartpad}
+/// This sample demonstrates how to use this [canPop] and [onPopInvoked] to
```
Comment (justinmc) [inline packages/flutter/lib/src/widgets/navigator.dart:639]: "this [canPop] and [onPopInvoked]" => "the [canPop] and [onPopInvoked] parameters"

File: examples/api/lib/widgets/page/page_can_pop.0.dart [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+        ),
+    ];
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return Navigator(
+      key: navigatorKey,
+      pages: _getPages(),
+      onDidRemovePage: (Page<Object?> page) {
```
Comment (justinmc) [inline examples/api/lib/widgets/page/page_can_pop.0.dart:105]: So this is called when a system back happens I guess?

File: packages/flutter/test/widgets/navigator_test.dart [hunk 10]
```diff
@@ -2894,8 +2894,8 @@ void main() {
         error.toStringDeep(),
         equalsIgnoringHashCodes(
           'FlutterError\n'
-          '   The Navigator.onPopPage must be provided to use the\n'
-          '   Navigator.pages API\n',
+          '   Either onDidRemovePage or onPopPage must be provided to use the\n'
+          '   Navigator.pages API but not both.\n',
```
Comment (justinmc) [inline packages/flutter/test/widgets/navigator_test.dart:2898]: Anything else that should be tested here besides the example tests?

**Review by justinmc — APPROVED** [review]
LGTM with nits :+1:

**Pull request merged**